gossipsub.go

```diff
@@ -384,9 +536,18 @@ func (gs *GossipSubRouter) Publish(from peer.ID, msg *pb.Message) {
 			continue
 		}
 
+		if gs.floodPublish && from == gs.p.host.ID() {
```

Comment:
so this is adding in the flood publish option, flooding IFF its from us.

Reply:
yes, otherwise it would be floodsub :p

Reply:
Do messages ever expire? If not, can't an attacker:

1. Collect a set of messages from us.
2. Wait for us to "forget" we sent the messages.
3. Feed us our own messages back.

Causing us to repeatedly flood our own messages out to the network (assuming they've collected enough of them)?

Reply:
We should drop our own messages if we receive them from the network (although I'm still concerned about the general replay issue).

Reply:
That's a good suggestion, we should never re-broadcast our own messages from the network.

Reply:
So the simplest ways to implement is to check if `ReceivedFrom` also matches our own host ID.
If it doesn't and it claims to be from us, we can drop it.

Reply:
Do we have any general protection against re-sending the same valid messages over and over in a cycle?

Reply:
There is the timecache, which marks it for 2 minutes (configuration variable).
Outside of the timecach window it's up to the application to detect stale/replayed messages at validation time.

Reply:
Ah, got it. That's probably as good as we can get for now.

Reply:
I have added the `ReceivedFrom` check; if a message has source as ourselves and was not received from ourselves, it is ignored and not forwarded.